bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerCreateOp.java

```diff
@@ -200,4 +209,233 @@ private void createComplete(int rc, LedgerHandle lh) {
+                LOG.error("invalid ackQuorumSize {} > writeQuorumSize {}", builderAckQuorumSize, builderWriteQuorumSize);
+                cb.createComplete(BKException.Code.IncorrectParameterException, null, ctx);
+                return false;
+            }
+
+            if (builderAckQuorumSize <= 0) {
+                LOG.error("invalid ackQuorumSize {} <= 0", builderAckQuorumSize);
+                cb.createComplete(BKException.Code.IncorrectParameterException, null, ctx);
+                return false;
+            }
+
+            if (builderPassword == null) {
+                LOG.error("invalid null password");
+                cb.createComplete(BKException.Code.IncorrectParameterException, null, ctx);
+                return false;
+            }
+
+            if (builderDigestType == null) {
+                LOG.error("invalid null digestType");
+                cb.createComplete(BKException.Code.IncorrectParameterException, null, ctx);
+                return false;
+            }
+
+            if (builderCustomMetadata == null) {
+                LOG.error("invalid null customMetadata");
+                cb.createComplete(BKException.Code.IncorrectParameterException, null, ctx);
+                return false;
+            }
+
+            return true;
+        }
+
+        @Override
+        public CompletableFuture<WriteHandle> execute() {
+            CompletableFuture<WriteHandle> counter = new CompletableFuture<>();
+            create(new SyncCreateCallback(), counter);
+            return counter;
+        }
+
+        private void create(CreateCallback cb, Object ctx) {
+            if (!validate(cb, ctx)) {
+                return;
+            }
+            LedgerCreateOp op = new LedgerCreateOp(bk, builderEnsembleSize,
+                builderWriteQuorumSize, builderAckQuorumSize, DigestType.fromApiDigestType(builderDigestType),
+                builderPassword, cb, ctx, builderCustomMetadata);
+            ReentrantReadWriteLock closeLock = bk.getCloseLock();
+            closeLock.readLock().lock();
+            try {
+                if (bk.isClosed()) {
+                    cb.createComplete(BKException.Code.ClientClosedException, null, ctx);
+                    return;
+                }
+                op.initiate();
+            } finally {
+                closeLock.readLock().unlock();
+            }
+        }
+    }
+
+    private static class CreateAdvBuilderImpl implements CreateAdvBuilder {
+
+        private Long builderLedgerId;
+        private final int builderEnsembleSize;
+        private final int builderAckQuorumSize;
+        private final int builderWriteQuorumSize;
+        private final byte[] builderPassword;
+        private final Map<String, byte[]> builderCustomMetadata;
+        private final org.apache.bookkeeper.client.api.DigestType builderDigestType;
+        private final BookKeeper bk;
+
+         private CreateAdvBuilderImpl(CreateBuilderImpl other) {
```

Comment:
do you really need reassign this? can't you just hold a reference to `CreateBuilderImpl`?

Reply:
done